File: Current/Product/Production/Common/Common.Tasks/LoopItems.vb
Option Explicit On
Option Strict On

Imports NAnt.Core
Imports NAnt.Core.Attributes

Public MustInherit Class LoopItems
    Inherits DataTypeBase
    Implements IEnumerable

    Public Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
        Return Me.GetStrings
    End Function

    Protected MustOverride Function GetStrings() As IEnumerator

    Public Overridable Sub Executing(ByVal item As String)

    End Sub
End Class

File: Product/Production/Nant/CIFactory.NAnt.Tasks/Types/LoopItems.cs
using System;
using System.Collections;
using NAnt.Core;

namespace CIFactory.NAnt.Types
{
    public abstract class LoopItems : DataTypeBase, IEnumerable
    {
        #region Public Methods

        public virtual void Executing(string item)
        {
        }

        public System.Collections.IEnumerator GetEnumerator()
        {
            return this.GetStrings();
        }

        #endregion

        #region Protected Methods

        protected abstract IEnumerator GetStrings();

        #endregion

    }
}